[BUG] fixes multivariate dataframe tsfile writes (#732)

hedeershowk opened the pull request:
#### Reference Issues/PRs
https://github.com/aeon-toolkit/aeon/issues/732

#### What does this implement/fix? Explain your changes.

Added appropriate parameters for the dataframe ts file writer to output the proper header.  Also one minor fix to `write_to_tsfile()` (see comment below).  

#### Did you add any tests for the change?

Yes in `aeon/datasets/tests/test_data_writers.py` I added `test_write_dataframe_to_ts_and_load`.  However, this test is currently failing, see comment below.

#### Any other comments?

* In `write_to_tsfile()` the initial check was just calling `isinstance(pd.Dataframe)` which I switched to `isinstance(X, pd.DataFrame)`.  It should have raised at some point though but maybe that function has never been called before without X being a `list` or `numpy.ndarray`? 
* I don't expect my current test to pass because the behavior of `load_from_tsfile()` hasn't been changed in this PR.  I wasn't sure if the idea was to keep `load_from_tsfile()` and `load_from_tsfile_to_dataframe()` as explicitly separate operations or if a user writing a dataframe to a a ts file should expect `load_from_tsfile()` to also return them a dataframe (right now it's np.ndarray).  I would personally expect a dataframe if I've written a dataframe out but I'm not sure whether that's the intention and if so how to store that information within the header (or elsewhere?) so the loader can be made aware.  Maybe a `@datatype` in the header?   

### PR checklist

<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.
- [ ] For new estimators, I've added the estimator to the online documentation.


<!--
Thanks for contributing!
-->

TonyBagnall commented:
hi, thanks for this, its very welcome. The dataframe loaders/writers are legacy really, back to the time when we stored collections in nested dataframes when each cell was a pd.Series. Good spot on the isinstance check, also found it here when looking to improve coverage. https://github.com/aeon-toolkit/aeon/pull/733. I am happy to leave 733 and update from yours once we get it in. I'll take a look today.

TonyBagnall commented on aeon/datasets/tests/test_data_writers.py:
```diff
@@ -98,3 +98,27 @@ def test_write_dataframe_to_ts():
+    # output the dataframe in a ts file
+    _write_dataframe_to_tsfile(
+        X=X,
+        path="./Temp/",
```
using "./Temp/" occasionally fails on CI, in #733 I switched to

    path = os.path.dirname(__file__) + "/Temp/"

which seems more reliable

hedeershowk replied:
Okay I'll update.  How do you feel about os.path.join instead of +?

TonyBagnall replied:
I think its good :)

TonyBagnall commented:
>I wasn't sure if the idea was to keep load_from_tsfile() and load_from_tsfile_to_dataframe()

hi @hedeershowk, just to clarify whats going on here, because its not necessarily intuitive and is a bit of a legacy to the origins of aeon  ....

load_from_tsfile() just loads data into either numpy (if series are equal length) or list of numpy (if series are unequal). These are the standard formats for the machine learning side (i.e. not forecasting): for equal length series stored as numpy is shape (n_cases, n_channels, n_timepoints). This structure is not amenable to standard pd.DataFrames, which in the simplest case for equal length with n_channels == 1 would store in shape (n_cases, n_timepoints) or (n_timepoints, n_cases), and if n_channels > 1 it gets complicated. You can use hierarchical/long formats,  but its messy.

Originally, some contributors really wanted to use data frames, so overcame these problems by storing data in a pd.DataFrame shape (n_cases, n_channels), with each entry a time series of length n_timepoints. This non standard usage tended to confuse everyone, so we have more recently attempted to implicitly deprecate it by removing usage. The x_to_dataframe functions are a legacy feature to load and read in this format. It means in recent refactoring I tried to ignore them, but you have found me out :)

So back to your PR. ITs useful, thanks, and good spot with the bug. I'd very much like to encourage your participation, we try to be as welcoming as possible. I will try take a look at how it fits in tomorrow, but I have quite a lot on, sorry if it takes me a little longer. 

If you want to  discuss further, please join our slack, as in #general or #contributors, or direct message me.

hedeershowk commented:
>  >I wasn't sure if the idea was to keep load_from_tsfile() and load_from_tsfile_to_dataframe()
> 
> hi @hedeershowk, just to clarify whats going on here, because its not necessarily intuitive and is a bit of a legacy to the origins of aeon  ....
> 
> load_from_tsfile() just loads data into either numpy (if series are equal length) or list of numpy (if series are unequal). These are the standard formats for the machine learning side (i.e. not forecasting): for equal length series stored as numpy is shape (n_cases, n_channels, n_timepoints). This structure is not amenable to standard pd.DataFrames, which in the simplest case for equal length with n_channels == 1 would store in shape (n_cases, n_timepoints) or (n_timepoints, n_cases), and if n_channels > 1 it gets complicated. You can use hierarchical/long formats,  but its messy.
> 
> Originally, some contributors really wanted to use data frames, so overcame these problems by storing data in a pd.DataFrame shape (n_cases, n_channels), with each entry a time series of length n_timepoints. This non standard usage tended to confuse everyone, so we have more recently attempted to implicitly deprecate it by removing usage. The x_to_dataframe functions are a legacy feature to load and read in this format. It means in recent refactoring I tried to ignore them, but you have found me out :)
> 
> So back to your PR. ITs useful, thanks, and good spot with the bug. I'd very much like to encourage your participation, we try to be as welcoming as possible. I will try take a look at how it fits in tomorrow, but I have quite a lot on, sorry if it takes me a little longer. 
> 
> If you want to  discuss further, please join our slack, as in #general or #contributors, or direct message me.    
> 
> 
>   

Ah okay thanks for the background information.  Makes sense and I can appreciate the desire to get things into data frames (in some cases) and the complexity of getting that done properly with the shape. 

I could update the test to more explicitly test the header construction the way you did in the bug report or just use the dataframe loader.  No rush here, just let me know what you'd prefer to get the PR how it can be most useful to you.

I've also joined the Slack (thanks) and look forward to participating and contributing more.

hedeershowk commented:
cleaned up that test and I switched to [tempfile](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory).  I was actually running into a bunch of file not found errors when running tests in this directory (even in the main branch).  Not sure why just running `pytest` from the module root didn't fail but when I explicitly ran pytest on `test_data_writers.py` I had a bunch of fails.  This PR fixes that.

hedeershowk commented:
hm haven't dug into it but looks like maybe the test fail in `test_proportion_train_in_param_finding()` is related to the shortcut in base when dealing with a single class?  I'll try to reproduce and resolve it locally.

hedeershowk commented:
Not clear how my changes could have an impact on the current failing test in CI?  I'll run locally to confirm and see if I can figure out the issue as well.

TonyBagnall reviewed (commented):
it looks good but I'm not sure about the changes to the Makefile @MatthewMiddlehurst can you take a look please?

TonyBagnall commented:
@hedeershowk could you update this with main (and resolve contributors conflict)? I think a lot of these changes were also in your other PR. Thanks, Tony
